core/options.c

```diff
@@ -618,57 +730,220 @@ check_param_bounds(uint *val, uint min, uint max, const char *name)
-#define DIFF_uint_size(value1,value2) ( value1 != value2 )
-#define DIFF_uint_time(value1,value2) ( value1 != value2 )
-#define DIFF_uint_addr(value1,value2) ( value1 != value2 )
-#define DIFF_pathstring_t(value1,value2) (strcmp(value1,value2))
-#define DIFF_liststring_t(value1,value2) (strcmp(value1,value2))
-
-#define COPY_bool(value1,value2) ( value1 = value2 )
-#define COPY_uint(value1,value2) ( value1 = value2 )
-#define COPY_uint_size(value1,value2) ( value1 = value2 )
-#define COPY_uint_time(value1,value2) ( value1 = value2 )
-#define COPY_uint_addr(value1,value2) ( value1 = value2 )
-#define COPY_pathstring_t(value1,value2) (strncpy(value1,value2,sizeof(value1)))
-#define COPY_liststring_t(value1,value2) (strncpy(value1,value2,sizeof(value1)))
+static void
+PRINT_STRING_uint_size(char *optionbuff, const void *val_ptr, const char *option)
+{
+    uint value = *(const uint*)val_ptr;
+    snprintf(optionbuff, MAX_OPTION_LENGTH, "-%s %d%s ", option,
+             (value % 1024 == 0 ? value/1024 : value),
+             (value % 1024 == 0 ? "K" : "B"));
+}
+static void
+PRINT_STRING_uint_time(char *optionbuff, const void *val_ptr, const char *option)
+{
+    uint value = *(const uint*)val_ptr;
+    snprintf(optionbuff, MAX_OPTION_LENGTH, "-%s %d ", option, value);
+}
+static void
+PRINT_STRING_uint_addr(char *optionbuff, const void *val_ptr, const char *option)
+{
+    ptr_uint_t value = *(const ptr_uint_t*)val_ptr;
+    snprintf(optionbuff, MAX_OPTION_LENGTH, "-%s "PFX" ", option, value);
+}
+static void
+PRINT_STRING_pathstring_t(char *optionbuff, const void *val_ptr, const char *option)
+{
+    snprintf(optionbuff, MAX_OPTION_LENGTH, "-%s '%s' ", option, (*(const pathstring_t*)val_ptr));
+}
+static void
+PRINT_STRING_liststring_t(char *optionbuff, const void *val_ptr, const char *option)
+{
+    snprintf(optionbuff, MAX_OPTION_LENGTH, "-%s '%s' ", option, (*(const liststring_t*)val_ptr));
+}
+
+static void
+print_fn(enum option_type_t type, char *optionbuff, const void *val_ptr,
+         const char *option)
+{
+    switch(type)
+    {
+      case OPTION_TYPE_bool:
+          PRINT_STRING_bool(optionbuff, val_ptr, option);
+          break;
+      case OPTION_TYPE_uint:
+          PRINT_STRING_uint(optionbuff, val_ptr, option);
+          break;
+      case OPTION_TYPE_uint_size:
+          PRINT_STRING_uint_size(optionbuff, val_ptr, option);
+          break;
+      case OPTION_TYPE_uint_time:
+          PRINT_STRING_uint_time(optionbuff, val_ptr, option);
+          break;
+      case OPTION_TYPE_uint_addr:
+          PRINT_STRING_uint_addr(optionbuff, val_ptr, option);
+          break;
+      case OPTION_TYPE_pathstring_t:
+          PRINT_STRING_pathstring_t(optionbuff, val_ptr, option);
+          break;
+      case OPTION_TYPE_liststring_t:
+          PRINT_STRING_liststring_t(optionbuff, val_ptr, option);
+          break;
+    }
+}
+
+static int
+DIFF_bool(const void *ptr1, const void *ptr2)
```

Comment:
Optional: drop to lower case (was upper b/c macro -- lower matches func style name) (ditto for others above)